pwiz_tools/Skyline/SettingsUI/Irt/CalibrateIrtDlg.cs

```diff
@@ -692,46 +693,67 @@ private static List<MeasuredPeptide> FindBestPeptides(SrmDocument doc, int pepti
-                for (int i = 0; i < peptideCount; i++)
-                {
-                    double targetRT = minRT + i * (gradientLength / (peptideCount - 1));
-                    for (int j = 0; j < docPeptides.Count; j++)
+                var resultsHandler = new MProphetResultsHandler(doc, model);
+                resultsHandler.ScoreFeatures(null, true);
+
+                var docTargets = docPeptides.ToDictionary(pep => pep.Target, pep => pep);
+                var docNodePeps = doc.Molecules.Where(nodePep => docTargets.ContainsKey(nodePep.ModifiedTarget)).ToArray();
+                var scoredPeptides = new List<ScoredPeptide>(docNodePeps.Length);
+                foreach (var nodePep in docNodePeps)
+                {
+                    var allStats = doc.MeasuredResults.MSDataFileInfos
+                        .Select(info => resultsHandler.GetPeakFeatureStatistics(nodePep.Id.GlobalIndex, info.FileId.GlobalIndex))
+                        .Where(stats => stats != null).ToArray();
+                    var value = float.MaxValue;
+                    if (allStats.Length > 0)
                     {
-                        if (j + 1 > docPeptides.Count - 1 ||
-                                Math.Abs(docPeptides[j].RetentionTime - targetRT) <
-                                Math.Abs(docPeptides[j + 1].RetentionTime - targetRT))
-                        {
-                            standardPeptides.Add(docPeptides[j]);
-                            docPeptides.RemoveAt(j);
-                            break;
-                        }
+                        value = model is MProphetPeakScoringModel
+                            ? allStats.Select(stats => stats.QValue.Value).Max()
+                            : -allStats.Select(stats => stats.BestScore).Min();
                     }
+                    scoredPeptides.Add(new ScoredPeptide(docTargets[nodePep.ModifiedTarget], value));
                 }
 
-                return standardPeptides;
+                var retentionTimes = scoredPeptides.Select(pep => pep.Peptide.RetentionTime).OrderBy(rt => rt).ToArray();
+                var minRt = retentionTimes.First();
+                var maxRt = retentionTimes.Last();
+                var rtRange = maxRt - minRt;
+
+                var buckets = PeptideBucket.BucketPeptides(scoredPeptides, new[]
+                {
+                    minRt + rtRange * 1 / 8,
+                    minRt + rtRange * 2 / 8,
+                    minRt + rtRange * 4 / 8,
+                    minRt + rtRange * 6 / 8,
+                    minRt + rtRange * 7 / 8,
+                    double.MaxValue
+                });
+                var endBuckets = new[] {buckets.First(), buckets.Last()};
+                var midBuckets = buckets.Skip(1).Take(buckets.Length - 2).ToArray();
+
+                var bestPeptides = new List<MeasuredPeptide>();
+                while (bestPeptides.Count < peptideCount)

```

Comment:
Maybe you should put something in this loop to make it so that you break out of the loop if all of the buckets are empty. I know that can't happen right now, but it's always good to be safe.